homeassistant/components/dynalite/bridge.py

```diff
@@ -1,118 +1,89 @@
 from dynalite_devices_lib import DynaliteDevices
 from dynalite_lib import CONF_ALL
 
-from homeassistant.const import CONF_HOST
 from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_send
 
-from .const import DATA_CONFIGS, DOMAIN, LOGGER
-from .light import DynaliteLight
-
-
-class BridgeError(Exception):
-    """Class to throw exceptions from DynaliteBridge."""
+from .const import CONF_NOWAIT, DATA_CONFIGS, DOMAIN, LOGGER
 
-    def __init__(self, message):
-        """Initialize the exception."""
-        super().__init__()
-        self.message = message
+CONNECT_TIMEOUT = 30
+CONNECT_INTERVAL = 1
 
 
 class DynaliteBridge:
     """Manages a single Dynalite bridge."""
 
-    def __init__(self, hass, config_entry):
+    def __init__(self, hass, host):
         """Initialize the system based on host parameter."""
-        self.config_entry = config_entry
         self.hass = hass
         self.area = {}
-        self.async_add_entities = None
-        self.waiting_entities = []
-        self.all_entities = {}
-        self.config = None
-        self.host = config_entry.data[CONF_HOST]
-        if self.host not in hass.data[DOMAIN][DATA_CONFIGS]:
-            LOGGER.info("invalid host - %s", self.host)
-            raise BridgeError(f"invalid host - {self.host}")
-        self.config = hass.data[DOMAIN][DATA_CONFIGS][self.host]
+        self.async_add_devices = None
+        self.waiting_devices = []
+        self.host = host
+        config = hass.data[DOMAIN][DATA_CONFIGS][self.host]
+        self.no_wait = config.get(CONF_NOWAIT, False)
         # Configure the dynalite devices
         self.dynalite_devices = DynaliteDevices(
-            config=self.config,
-            newDeviceFunc=self.add_devices,
+            config=config,
+            newDeviceFunc=self.add_devices_when_registered,
             updateDeviceFunc=self.update_device,
         )
 
     async def async_setup(self, tries=0):
         """Set up a Dynalite bridge."""
         # Configure the dynalite devices
-        await self.dynalite_devices.async_setup()
-
-        self.hass.async_create_task(
-            self.hass.config_entries.async_forward_entry_setup(
-                self.config_entry, "light"
-            )
-        )
-
-        return True
-
-    @callback
-    def add_devices(self, devices):
-        """Call when devices should be added to home assistant."""
-        added_entities = []
-
-        for device in devices:
-            if device.category == "light":
-                entity = DynaliteLight(device, self)
-            else:
-                LOGGER.debug("Illegal device category %s", device.category)
-                continue
-            added_entities.append(entity)
-            self.all_entities[entity.unique_id] = entity
-
-        if added_entities:
-            self.add_entities_when_registered(added_entities)
+        if not await self.dynalite_devices.async_setup():
+            return False
```

Comment:
If we can distinguish between bad set up and the gateway is down, raise two different errors here based on the case. If it's a non recoverable error we want to fail the entry setup. If it's a recoverable error we want to try to set up the entry again later by raising `ConfigEntryNotReady`.